### ISSUE TITLE
Accessing Request in Apollo Server 2

### ISSUE THREAD

**Issue by tadeolinco** [apollographql/apollo-server#1066]
Is there any way of accessing the request object from the underlying express app in Apollo Server 2 beta? Much like this one #420.

**Comment by eugene1g** [comment]
The `context` configuration parameter can be an object, or a function that returns the context object. This function gets the HTTP request as a parameter and could be defined like so -
```
const myServer =  new ApolloServer({
    schema: executableSchema,
    context: ({req}) => ({ user: req.session.user })
  });
```

**Comment by evans** [comment]
The context as a function is a perfect way to do this. Thanks @eugene1g for the code snippet!

It would be awesome if you could open a PR to the `version-2` branch and add the context to [the api reference](https://github.com/apollographql/apollo-server/blob/version-2/docs/source/api/apollo-server.md#apolloserver)

**Comment by darkadept** [comment]
@eugene1g Can context be a function that returns a Promise like you show above?

I'm running Apollo Server 2.0.0-beta.3 and my resolvers are receiving an unresolved Promise in context param.

My context function looks just like yours.

**Comment by eugene1g** [comment]
@darkadept Nope! I just tested `2.0.0-beta.3`, and it no longer supports async context. Have to stay to `beta.2` for now 

/cc @evans

**Comment by eugene1g** [comment]
@darkadept I looked through the code and looks like there is regression in how `context` gets passed through the layers. I can't tell if it's been corrected in master as there is quite a bit of refactoring going on.

However, there seems to be a hacky workaround: it will work if you wrap your real async function in another vanilla function like so -
```js
const generateContext = async ({req}) => { 
        const data = await someStuff(); 
        return {
            token: data.stuff, 
            user: req.session.user 
        } 
}
const myServer =  new ApolloServer({
    schema: executableSchema,
    context: params => () => generateContext(params);
  });
```

This "works" in a sense is that it gives you async context generators in `beta.3` but this behavior is not standard and well may break in the next beta!

**Comment by darkadept** [comment]
@eugene1g Oh wow. That works! Thanks.

What's the roadmap for this context function though? Is it supposed to support async context functions in the future? I just realized this should probably be it's own issue. Should I open an issue for this?

**Comment by evans** [comment]
@eugene1g Definitely a regression! I'll add a test and publish another beta

**Comment by eugene1g** [comment]
Confirming this has been fixed `2.0.0-beta.3` and `context` handles async functions. Thanks!

**Comment by Lahori-Jawan** [comment]
@eugene1g can you help me on this [issue](https://github.com/apollographql/apollo-server/issues/1791)? your approach seems a better alternative but doesn't work for me.
